ipld/read.go

```diff
@@ -18,190 +17,177 @@ import (
+// fillQuadrant fetches 1/4 of shares for the given root
+func fillQuadrant(
+	ctx context.Context,
+	quadrant *quadrant,
+	dag ipld.NodeGetter,
+	dataSquare [][]byte,
+) error {
+	errGroup, ctx := errgroup.WithContext(ctx)
+	for i := 0; i < len(quadrant.rootCids); i++ {
+		i := i
+		errGroup.Go(func() error {
+			leaves, err := GetSubtreeLeaves(
+				ctx,
+				quadrant.rootCids[i],
+				dag,
+				quadrant.isLeftSubtree,
+				uint32(len(quadrant.rootCids)/2),
+			)
+			if err != nil {
+				return err
+			}
+			quadrantWidth := len(quadrant.rootCids)
+			for leafIdx, leaf := range leaves {
+				// shares from the right subtree should be
+				// inserted after all shares from the left subtree
+				if !quadrant.isLeftSubtree {
+					leafIdx += len(leaves)
+				}
+				shareData := leaf.RawData()[1:]
+				// dataSquare represents a single dimensional slice
+				// of 4 quadrants. The representation of quadrants in dataSquare will be
+				// | 1 | | 2 | | 3 | | 4 |
+				// position is calculated by offsetting the index to respective quadrant
+				position := ((i + quadrant.from) * quadrantWidth * 2) + leafIdx
+				dataSquare[position] = shareData[NamespaceSize:]
 
-	return &shareCounter{
-		shares:          make(map[index][]byte),
-		edsWidth:        edsWidth,
-		minSharesNeeded: minSharesNeeded,
-		shareChan:       make(chan indexedShare, 512),
-		errc:            make(chan error, 1),
-		ctx:             ctx,
-		cancel:          cancel,
+			}
+			return err
+		})
 	}
+	return errGroup.Wait()
 }
 
-// retrieveLeaf uses GetLeafData to fetch a single leaf and counts that leaf
-func (sc *shareCounter) retrieveShare(
-	rootCid cid.Cid,
-	isRow bool,
-	axisIdx uint32,
-	idx uint32,
+// GetSubtreeLeaves returns only one subtree - left or right
+func GetSubtreeLeaves(
+	ctx context.Context,
+	root cid.Cid,
 	dag ipld.NodeGetter,
-) {
-	data, err := GetLeafData(sc.ctx, rootCid, idx, sc.edsWidth, dag)
+	isLeftSubtree bool,
+	treeSize uint32,
+) ([]ipld.Node, error) {
+	nd, err := dag.Get(ctx, root)
 	if err != nil {
-		select {
-		case <-sc.ctx.Done():
-		case sc.errc <- err:
-		}
-	}
-
-	if len(data) < plugin.ShareSize {
-		return
+		return nil, err
 	}
-
-	// switch the row and col indexes if needed
-	rowIdx := idx
-	colIdx := axisIdx
-	if isRow {
-		rowIdx = axisIdx
-		colIdx = idx
+	links := nd.Links()
+	subtreeRootHash := links[0].Cid
+	if !isLeftSubtree && len(links) > 1 {
+		subtreeRootHash = links[1].Cid
 	}
 
-	select {
-	case <-sc.ctx.Done():
-	case sc.shareChan <- indexedShare{data: data[NamespaceSize:], index: index{row: rowIdx, col: colIdx}}:
-	}
+	return getLeaves(ctx, dag, subtreeRootHash, make([]ipld.Node, 0, treeSize))
 }
 
-// wait until enough data has been collected, the timeout has been reached, or
-// too many errors are encountered
-func (sc *shareCounter) wait() error {
-	defer sc.cancel()
+func GetLeaves(ctx context.Context, dag ipld.NodeGetter, root cid.Cid, size uint32) ([]ipld.Node, error) {
```

Comment:
Can this exported function be documented? Also can it have a unit test as well?

Reply:
removed